### ISSUE TITLE
Inconsistency between appendQuery and mergeQuery in an edge case

### ISSUE THREAD

**Issue by artem-tim** [thephpleague/uri-components#30]
### Bug Report

_(Fill in the relevant information below to help triage your issue.)_

| Information | Description |
|--------------|---------|
| Version (league/uri) | 6.0 |
| Version (league/uri-components) | 2.0 |
| PHP version | 7.2 |
| OS Platform | 18.04.1-Ubuntu |

#### Summary

`mergeQuery` is not consistent with `appendQuery` when input URI string has empty query but contains `?` character, neither with itself in case when input URI string does not contain `?` character

#### Standalone code, or other way to reproduce the problem

```
{   
    $url = Uri::createFromString("www.host.com/path");
    $query = $url->getQuery(); // Empty string
    $newUrlAppend = UriModifier::appendQuery($url, "foo=baz");  // www.host.com/path?foo=baz
    $newUrlMerge = UriModifier::mergeQuery($url, "foo=baz");    // www.host.com/path?foo=baz
}
{   
    $url = Uri::createFromString("www.host.com/path?");
    $query = $url->getQuery(); // Empty string
    $newUrlAppend = UriModifier::appendQuery($url, "foo=baz");  // www.host.com/path?foo=baz
    $newUrlMerge = UriModifier::mergeQuery($url, "foo=baz");    // www.host.com/path?&foo=baz
}

```

#### Expected result

`$newUrlMerge` of the second block should print `www.host.com/path?foo=baz`

#### Actual result

`$newUrlMerge` of the second block prints `www.host.com/path?&foo=baz`

**Comment by nyamsprod** [comment]
@artem-tim I've looked into your report and this is more of a documentation issue than a bug per se  but then again it might be perceived as one 😢 

`Query::merge` adds or replaces pairs from the subject query with pairs from the input query

```php
<?php

$subjectQuery = new Query('');
$inputQuery = new Query('foo=bar');
echo $subjectQuery; //return '' the empty string
echo $subjectQuery->merge($inputQuery)->getContent(); // return '&foo=bar'
````
Here, the input query did not contained a pair with an empty key, hence **the pair with an empty key `['' => null]` is not removed or replace.**

`Query::append` adds the input query at the end of the subject query.

```php
<?php

$subjectQuery = new Query('');
$inputQuery = new Query('foo=bar');
echo $subjectQuery; //return '' the empty string
echo $subjectQuery->append($inputQuery)->getContent(); // return 'foo=bar'
````

When appending to avoid the multiplication of `&` where queries intersect the `append` method strip them prior to `glue` both strings together so **the pair with an empty key `['' => null]` is removed** resulting in what you observe.

This behavior is expected and covered by tests in the test suite see

https://github.com/thephpleague/uri-components/blob/0f1568baca257418ae2419205e9472a41d944c40/tests/Components/QueryTest.php#L228-L259

I hope this clarify what you've come accrossed.

**Comment by artem-tim** [comment]
@nyamsprod Thank you for taking time looking into this.
I understand why it **is** happening, but not why it **has to be** happening.

On my side, I couldn't find any RFC line about this, and all validators I tried accept this.
So, apart from "not feeling right" (and it **feels like** the pair `['' => null] ` should be filtered out from query param array), there is indeed no documentation to classify this behavior as a bug. However, in the same way there is no reason to not making `merge` work better, is there ?

This extends to cases where there are 2 or more `&` symbols between params, they are preserved as-is for `merge` but properly cleaned by `append`. For example :
```
$url = Uri::createFromString("/?a=1&b=2&&&");
$urlAppend = UriModifier::appendQuery($url, "c=3"); // => /?a=1&b=2&c=3
$urlMerge = UriModifier::mergeQuery($url, "c=3"); // => /?a=1&b=2&&&&c=3
```
`append` has the desired *query cleaning* behavior (as a secondary effect), but `merge` being a different, equally (even more) useful feature, does not have it.
In conclusion it's not a bug, but more like a missing feature :)

**Comment by nyamsprod** [comment]
> append has the desired query cleaning behavior (as a secondary effect), but merge being a different, equally (even more) useful feature, does not have it.

there's a `Query::withoutEmptyPairs` method for the clean up You can use it on the resulting merged query. And, IMHO, the one method at fault is more `append` than `merge`.

> On my side, I couldn't find any RFC line about this, and all validators I tried accept this.

That's normal because there is no actual RFC to describe the query string and how it should be manipulated as it is an implementation details which depends on the URL scheme. The only RFC-ish document is the one from the WHATWG but it describes `Query::appendTo`.

**Comment by artem-tim** [comment]
Thanks. Please judge whether to deal with this inconsistency or close the issue. I got my answer :)

**Comment by nyamsprod** [comment]
@artem-tim If I update the documentation to hint that for the clean up you *may* want to use the `Query::withoutEmptyPairs`  after `Query::merge` is it OK for you ? The reason I ask is because we can not satisfy everyone and having merge staying strict as it is now was more useful in my mind so giving this hint might help other 🤔

**Comment by artem-tim** [comment]
@nyamsprod Yeah I think it's best to mention that. I don't quite see the point of keeping empty pairs in the first place, so I'm guessing that's what people will end up doing most of the time with `merge`.
